pkg/controllers/management/template/template_cluster_catalog.go

```diff
@@ -0,0 +1,190 @@
+package template
+
+import (
+	"reflect"
+	"sort"
+
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	log "github.com/sirupsen/logrus"
+
+	"k8s.io/api/rbac/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+)
+
+var rolesWithAccessToClusterCatalogTemplates = map[string]bool{
+	"cluster-owner":          true,
+	"cluster-member":         true,
+	"clustercatalogs-manage": true,
+	"clustercatalogs-view":   true,
+}
+
+func (tm *RBACTemplateManager) reconcileClusterCatalog(clusterID string) error {
+	var clusterTemplates, clusterTemplateVersions []string
+	// Get all cluster catalogs created in this cluster
+	clusterCatalogs, err := tm.clusterCatalogLister.List(clusterID, labels.NewSelector())
+	if err != nil {
+		log.Errorf("Error while listing cluster catalogs: %v", err)
+		return err
+	}
+	// Get all templates and template versions for each of these cluster catalogs.
+	// Templates for cluster catalog are created with label "clusterId-catalogName=catalogName"
+	for _, clusterCatalog := range clusterCatalogs {
+		c := clusterID + "-" + clusterCatalog.Name
+		r, err := labels.NewRequirement(c, selection.Equals, []string{clusterCatalog.Name})
+		if err != nil {
+			return err
+		}
+		//get templates, template versions for this cluster catalog
+		templates, templateVersions, err := tm.getTemplateAndTemplateVersions(r)
+		// add it to the cluster's templates templateVersions list
+		clusterTemplates = append(clusterTemplates, templates...)
+		clusterTemplateVersions = append(clusterTemplateVersions, templateVersions...)
+	}
+	// have them sorted for deepEqual calls later
+	sort.Strings(clusterTemplates)
+	sort.Strings(clusterTemplateVersions)
+	newRules := []v1.PolicyRule{
+		{
+			APIGroups:     []string{"management.cattle.io"},
+			Resources:     []string{"templates"},
+			ResourceNames: clusterTemplates,
+			Verbs:         []string{"get", "list", "watch"},
+		},
+		{
+			APIGroups:     []string{"management.cattle.io"},
+			Resources:     []string{"templateversions"},
+			ResourceNames: clusterTemplateVersions,
+			Verbs:         []string{"get", "list", "watch"},
+		},
+	}
+
+	// Add these templates/templateversions to the clusterRole for this cluster's templates/templateVersions
+	// check if the clusterRole exists
+	cRole, err := tm.clusterRoleClient.Get("cluster-"+clusterID+clusterRoleSuffix, metav1.GetOptions{})
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+		// error is IsNotFound, so create it
+		newCR := &v1.ClusterRole{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "cluster-" + clusterID + clusterRoleSuffix,
```

Comment:
this should be defined at the top of the method, or even in the separate method as it's being referred multiple times.